### PR TITLE
Feat/notifications

### PR DESCRIPTION
**Pull request by ProggerOleg**
## Describe your changes

- Created notifications module and logic
- Made endpoint to get notifications for certain user ordered by desc
- Implemented notification creation for different events and appointment statuses
- Fixed bug, now user get's his money back if appointment was rejected

Screenshot from 2024-01-10 13-21-56

## Issue ticket code (and/or) and link

- [Link to JIRA ticket](https://homecareaid.atlassian.net/browse/CC-282)

### **General**

- [x] Assigned myself to the PR
- [x] Assigned the appropriate labels to the PR
- [x] Assigned the appropriate reviewers to the PR
- [x] Updated the documentation
- [x] Performed a self-review of my code
- [x] Types for input and output parameters
- [x] Don't have "any" on my code
- [x] Used the try/catch pattern for error handling
- [x] Don't have magic numbers
- [x] Compare only with constants not with strings
- [x] No ternary operator inside the ternary operator
- [x] Don't have commented code
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file.
- [x] Used camelCase for variables and functions
- [ ] Date and time formats are on the constants
- [x] Functions are public only if it's used outside the class
- [x] No hardcoded values
- [ ] covered by tests
- [x] Check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/).

### Backend

- [x] Swagger documentation updated
- [x] Database requests are optimized and not redundant
- [ ] Unit tests written
- [ ] use ConfigService instead of process.env
- [x] use transactions if there is a call chain that mutates data in different tables
- [ ] use @index decorator for frequently requested data

### CODE REVIEW

File: src/modules/appointment/appointment.service.ts [hunk 1]
```diff
@@ -479,7 +483,65 @@ export class AppointmentService {
         })
         .execute();
 
+      const singleAppointment = await this.findOneById(appointmentId);
+
       if (appointment.status) {
+        if (appointment.status === AppointmentStatus.Rejected) {
+          if (appointmentToUpdate.status === AppointmentStatus.Pending) {
+            if (role === UserRole.Caregiver) {
```
Comment (Lilly-nn) [inline src/modules/appointment/appointment.service.ts]: I feel that there too many if statements here, can't we refactor this part a little bit to make it more readable and concise?

File: src/modules/activity-log/activity-log.service.ts [hunk 2]
```diff
@@ -45,6 +83,31 @@ export class ActivityLogService {
...
+          activityLog.appointment.userId,
+        );
+      } else if (updateActivityLogDto.status === ActivityLogStatus.Pending) {
+        await this.notificationService.createNotification(
+          activityLog.appointment.userId,
+          activityLog.appointment.id,
+          NotificationMessage.ActivityLogReview,
+          activityLog.appointment.caregiverInfo.user.id,
+        );
+      }
```
Comment (Simpsonkk) [inline src/modules/activity-log/activity-log.service.ts]: Is it possible to get rid of duplication and call createNotification only once?

Comment (ProggerOleg) [reply src/modules/activity-log/activity-log.service.ts]: Agree, fixed

File: src/modules/appointment/appointment.service.ts (hunk 1 above)
Comment (ProggerOleg) [reply src/modules/appointment/appointment.service.ts]: This logic is quite complicated but i will try

**Review by Lilly-nn — APPROVED** [review]

**Review by Simpsonkk — APPROVED** [review]

**Pull request merged**